src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.cs

```diff
@@ -541,31 +555,34 @@ private int WriteFrameToSendBuffer(MessageOpcode opcode, bool endOfMessage, Read
             {
                 // The server doesn't send a mask, so the mask offset returned by WriteHeader
                 // is actually the end of the header.
-                headerLength = WriteHeader(opcode, _sendBuffer, payloadBuffer, endOfMessage, useMask: false);
+                headerLength = WriteHeader(opcode, _sendBuffer, payloadBuffer, endOfMessage, useMask: false, compressed: _inflater is not null);
             }
             else
             {
                 // We need to know where the mask starts so that we can use the mask to manipulate the payload data,
                 // and we need to know the total length for sending it on the wire.
-                maskOffset = WriteHeader(opcode, _sendBuffer, payloadBuffer, endOfMessage, useMask: true);
+                maskOffset = WriteHeader(opcode, _sendBuffer, payloadBuffer, endOfMessage, useMask: true, compressed: _inflater is not null);
                 headerLength = maskOffset.GetValueOrDefault() + MaskLength;
             }
 
             // Write the payload
-            if (payloadBuffer.Length > 0)
+            if (!payloadBuffer.IsEmpty)
```

Comment:
`Length` is preferred to `IsEmpty` with spans when the expectation is that the span points to a valid buffer.